Document python requirement for tox/mypy & remove basepython from conf (#2644)
sqlfluff/sqlfluff: CONTRIBUTING.md

### Developing and Running SQLFluff Locally

#### Requirements

The simplest way to set up a development environment is to use `tox`.
First ensure that you have tox installed:
```shell
python3.8 -m pip install -U tox
```

> **IMPORTANT:** `tox` must be installed using Python 3.8 or newer. Otherwise `mypy` checks may fail. Depending on your setup and personal preference, replace `python3.8` as necessary, but make sure that the python version is new enough.

Reviewer: ```suggestion
**IMPORTANT:** `tox` must be installed with a minimum of Python 3.8 as the `mypy` checks are incompatible with 3.7. Those using newer versions of Python may replace `python3.8` as necessary (the test suite runs primarily under 3.10 for example).
```

> <!-- Hopefully when tox 4 is released it will be possible to set `basepython = python3.10, python3.9, python3.8` to enforce a working version for running mypy checks. For now it's only possible to set it to a single version which is too restrictive. On the other hand for now there is a requirement for contributors to install tox with a new enough Python version. -->

Reviewer: I think this is a little longwinded and potentially confusing. I note you commented it out, but I for one often read CONTRIBUTING.md and the like in my IDE where it won’t be hidden.

WDYT about this?:

```suggestion
Note: Unfortunately tox does not currently support setting just a minimum Python version (though this may be be coming in tox 4!).
```

Author: That's well put. Thanks.

#### Creating a virtual environment

> A virtual environment can then be created and activated by running (check the [requirements](#requirements) before running this):

Reviewer: Not sure this is needed and reads a little confusing. I think the `Then` covers it.

```suggestion
A virtual environment can then be created and activated by running:
```

Author: I wanted to be extra sure. If someone just jumps to this section and happens to have tox installed previously, they might miss the requirement. Are you ok with keeping the pointer?

Reviewer: Yup. OK keeping them all as you prefer.

### Testing

> To test locally, SQLFluff uses `tox` (check the [requirements](#requirements)!). The test suite can be run via:

Reviewer: Same feedback here. I know there’s a chance people can miss the early point but personally still find it a little overkill.

```suggestion
To test locally, SQLFluff uses `tox`. The test suite can be run via:
```

### Pre-Commit Config

> For development convenience we also provide a `.pre-commit-config.yaml` file to allow the user to install a selection of pre-commit hooks by running (check the [requirements](#requirements) before running this):

Reviewer: And here. Do think we’ve gone from too little guidance here to too much. But maybe you feel strong about leaving some of these in place since I k ow you say you wasted a good bit of time on this?

```suggestion
For development convenience we also provide a `.pre-commit-config.yaml` file to allow the user to install a selection of pre-commit hooks by running:
```

Author: Out of the cases discussed, this is maybe the most important to keep the pointer. Because this is so far away from the main tox stuff, I can easily imagine someone jumping here just for the sake of installing pre-commit tools and again if that someone already has tox installed, it's not obvious that it could be wrong even though `tox -e pre-commit -- install` seems to succeed.

Author: To make things worse, if you already ran this before you realized you need to change your tox installation, it's not trivial to realize that you also need to wipe out the `.tox/pre-commit` by hand and run `tox -e pre-commit -- install` again. Better safe than sorry, it's a pity that we can't enforce these programmatically.
